### PR TITLE
Preserve domain slot ordering and multistrings while dumping

### PR DESCRIPTION
**Pull request by degiz**
Fix for https://github.com/RasaHQ/rasa-x/issues/3900 and https://github.com/RasaHQ/rasa-x-demo/issues/27

**Proposed changes**:
- Preserve ordering of domain slots while dumping
- Preserve multi line strings in responses `text` key

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/shared/core/domain.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import json
 import logging
 import os
+import textwrap
```
Comment (wochinge) [inline rasa/shared/core/domain.py:6]:
```suggestion
```

File: rasa/shared/core/domain.py [hunk 2]
```diff
@@ -1045,12 +1047,39 @@ def as_dict(self) -> Dict[Text, Any]:
...
             KEY_ACTIONS: self._custom_actions,  # class names of the actions
             KEY_FORMS: self.forms,
             KEY_E2E_ACTIONS: self.action_texts,
         }
 
+    @staticmethod
+    def get_responses_with_multilines(
+        responses: Dict[Text, List[Dict[Text, Any]]]
+    ) -> Dict[Text, List[Dict[Text, Any]]]:
+        """Return `responses` with preserved multilines in the `text` key.
```
Comment (wochinge) [inline rasa/shared/core/domain.py:1060]:
```suggestion
        """Returns `responses` with preserved multilines in the `text` key.
```

File: rasa/shared/core/domain.py [hunk 3]
```diff
@@ -1045,12 +1047,39 @@ def as_dict(self) -> Dict[Text, Any]:
             KEY_INTENTS: self._transform_intents_for_file(),
             KEY_ENTITIES: self._transform_entities_for_file(),
             KEY_SLOTS: self._slot_definitions(),
-            KEY_RESPONSES: self.templates,
+            KEY_RESPONSES: self.get_responses_with_multilines(self.templates),
```
Comment (wochinge) [inline rasa/shared/core/domain.py:1050]: I think we should call this in `as_yaml`. What do you think? Feels a bit weird in the regular `as_dict` function which is not necessarily used for dumping.

File: tests/shared/core/test_domain.py [hunk 4]
```diff
@@ -1142,3 +1145,63 @@ def test_valid_slot_mappings(domain_as_dict: Dict[Text, Any]):
 def test_form_invalid_mappings(domain_as_dict: Dict[Text, Any]):
     with pytest.raises(InvalidDomain):
         Domain.from_dict(domain_as_dict)
+
+
+def test_slot_order_is_preserved():
```
Comment (wochinge) [inline tests/shared/core/test_domain.py:1150]: Can we please also add a test where we merge domains and then make sure that the slot order still makes sense? Merging is usually the part where things go wrong in this module 🙈

**Review by wochinge — COMMENTED** [review]
Nice, concise fix! 

Should this rather go into `2.1.x`?

File: rasa/shared/core/domain.py (hunk 2 above)
Comment (degiz) [reply rasa/shared/core/domain.py:1060]: I need to make sure to read our style guide  every two weeks 🤦

**Review by wochinge — APPROVED** [review]

**Pull request merged**